db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -1177,7 +1177,11 @@ Status DBImpl::CompactRangeInternal(const CompactRangeOptions& options,
           break;
         }
         if (output_level == ColumnFamilyData::kCompactToBaseLevel) {
-          final_output_level = cfd->NumberLevels() - 1;
```

Comment:
The old code assumed the final output level of a `CompactRange()` with `level_compaction_dynamic_level_bytes==true` would be the last level. That assumption seems reasonable because dynamic leveled compaction is supposed to fill in levels starting with the last level, and `CompactRange()` is supposed to compact down to the last level containing any data (or last level if all of L1+ are empty). 

However as you found there can be edge cases where the final output level is not the last level. In the unit test it looks like the cause was a DB that already had data in L1+ reopened with `num_levels_new > num_levels_old` (`num_levels_old` is the pre-reopen value of `num_levels`, and `num_levels_new` is the post-reopen value). Then the target level of the post-reopen `CompactRange()` is `num_levels_old-1`, which is not the last level anymore. Another scenario it happens, often surprising/inconveniencing users, is after enabling `level_compaction_dynamic_level_bytes`. That makes the migration to enable `level_compaction_dynamic_level_bytes` difficult as even compacting the whole DB through `CompactRange()` with nullptr endpoints is not enough.

In my opinion, `CompactRange()` should move the data to the ideal level according to the current config, i.e., not stop at empty level when `level_compaction_dynamic_level_bytes==true`. I think the type of change in this PR adds complexity and has race conditions (base level can change after `RunManualCompaction()`). If there's a way to do it without the race conditions it might be worthwhile as a stop-gap solution. The conceptual simplicity of the old code will be missed (by me), though.

Reply:
> CompactRange() is supposed to compact down to the last level containing any data

I think the current behavior is to compact down to the lowest level which contains any data *that overlaps with the key range passed into `CompactRange()`*. So there are more cases where the old code would move wrong the wrong level during refit level.